ledger/complete/compactor.go

```diff
@@ -0,0 +1,407 @@
+	if err != nil {
+		return fmt.Errorf("error serializing checkpoint (%d): %w", checkpointNum, err)
+	}
+
+	duration := time.Since(startTime)
+	logger.Info().Float64("total_time_s", duration.Seconds()).Msgf("created checkpoint %d", checkpointNum)
+
+	return nil
+}
+
+func cleanupCheckpoints(checkpointer *realWAL.Checkpointer, checkpointsToKeep int) error {
+	// Don't list checkpoints if we keep them all
+	if checkpointsToKeep == 0 {
+		return nil
+	}
+	checkpoints, err := checkpointer.Checkpoints()
+	if err != nil {
+		return fmt.Errorf("cannot list checkpoints: %w", err)
+	}
+	if len(checkpoints) > int(checkpointsToKeep) {
+		// if condition guarantees this never fails
+		checkpointsToRemove := checkpoints[:len(checkpoints)-int(checkpointsToKeep)]
+
+		for _, checkpoint := range checkpointsToRemove {
+			err := checkpointer.RemoveCheckpoint(checkpoint)
+			if err != nil {
+				return fmt.Errorf("cannot remove checkpoint %d: %w", checkpoint, err)
+			}
+		}
+	}
+	return nil
+}
+
+// processTrieUpdate writes trie update to WAL, updates activeSegmentNum,
+// and gets tries from checkpointQueue for checkpointing if needed.
+// It also sends WAL update result and waits for trie update completion.
+func (c *Compactor) processTrieUpdate(
+	update *WALTrieUpdate,
+	checkpointQueue *CheckpointQueue,
+	activeSegmentNum int,
+	nextCheckpointNum int,
+) (
+	_activeSegmentNum int,
+	checkpointNum int,
+	checkpointTries []*trie.MTrie,
+) {
+
+	// RecordUpdate returns the segment number the record was written to.
+	// Returned segment number (>= 0) can be
+	// - the same as previous segment number (same segment), or
+	// - incremented by 1 from previous segment number (new segment)
+	segmentNum, skipped, updateErr := c.wal.RecordUpdate(update.Update)
+
+	// Send result of WAL update
+	update.ResultCh <- updateErr
+
+	// This ensures that updated trie matches WAL update.
+	defer func() {
+		// Wait for updated trie
+		trie := <-update.TrieCh
+		if trie == nil {
+			c.logger.Error().Msg("compactor failed to get updated trie")
+			return
+		}
+
+		checkpointQueue.Push(trie)
+	}()
+
+	if activeSegmentNum == -1 {
+		// Recover from failure to get active segment number at initialization.
+		return segmentNum, -1, nil
+	}
+
+	if updateErr != nil || skipped || segmentNum == activeSegmentNum {
+		return activeSegmentNum, -1, nil
+	}
+
+	// In the remaining code: segmentNum > activeSegmentNum
+
+	// active segment is finalized.
+
+	// Check new segment number is incremented by 1
+	if segmentNum != activeSegmentNum+1 {
+		c.logger.Error().Msg(fmt.Sprintf("compactor got unexpected new segment numer %d, want %d", segmentNum, activeSegmentNum+1))
+	}
+
+	// Update activeSegmentNum
+	prevSegmentNum := activeSegmentNum
+	activeSegmentNum = segmentNum
+
+	if nextCheckpointNum > prevSegmentNum {
+		// Not enough segments for checkpointing
+		return activeSegmentNum, -1, nil
+	}
+
+	// In the remaining code: nextCheckpointNum == prevSegmentNum
```

Comment:
Is it possible that `prevSegmentNum > nextCheckpointNum`? Especially in the case when checkpoint triggers again before the previous checkpoint finishes? Just to double check.

If this is possible, we might want to use `prevSegmentNum` to trigger checkpoint. Either way, I think it's always safer to use `prevSegmentNum` than `nextCheckpointNum`.

Reply:
> Is it possible that prevSegmentNum > nextCheckpointNum? 

No.  It is an invariant that `nextCheckpointNum >= activeSegmentNum`.  Because `prevSegmentNum` is initialized to `activeSegmentNum` when active segment is finalized and becomes previous segment, therefore `nextCheckpointNum >= prevSegmentNum`.

> Especially in the case when checkpoint triggers again before the previous checkpoint finishes?

When checkpoint triggers again before the previous checkpoint finishes, we fail to acquire semaphore and this block is executed so that checkpointing is retried when active segment is finalized.
https://github.com/onflow/flow-go/blob/240520d97949a6aecf4faa8f8ae5b1324822d139/ledger/complete/compactor.go#L225-L226